include/libvfio-user.h

```diff
@@ -268,33 +246,106 @@ typedef enum {
 } vfu_pci_type_t;
 
 /**
- * Setup PCI configuration space header data. This function must be called only
- * once per libvfio-user context.
+ * Initialize the context for a PCI device. Returns an initialized buffer for a
+ * type 0 PCI configuration space of the needed size.
+ *
+ * This buffer consists of the standard 64-byte PCI header and the remaining
+ * space depending on @pci_type, either PCI_CFG_SPACE_SIZE or
+ * PCI_CFG_SPACE_EXP_SIZE.
  *
  * @vfu_ctx: the libvfio-user context
- * @id: Device and vendor ID
- * @ss: Subsystem vendor and device ID
- * @cc: Class code
  * @pci_type: PCI type (convention PCI, PCI-X mode 1, PCI-X mode2, PCI-Express)
  * @revision: PCI/PCI-X/PCIe revision
+ */
+vfu_pci_hdr_t *
+vfu_pci_init(vfu_ctx_t *vfu_ctx, vfu_pci_type_t pci_type, int revision UNUSED);
+
+/*
+ * FIXME
+ */
+void
+vfu_pci_set_id(vfu_ctx_t *vfu_ctx, uint16_t vid, uint16_t did,
+               uint16_t ssid, uint16_t ssvid);
+
+/*
+ * FIXME
+ */
+void
+vfu_pci_set_class(vfu_ctx_t *vfu_ctx, uint8_t base, uint8_t sub, uint8_t pi);
+
+// FIXME: what else do we need? multi-function flag in "header type" ? rid?
+
+/**
+ * Add a PCI capability to PCI config space.
+ *
+ * Certain standard capabilities are handled entirely within the library:
+ *
+ * PCI_CAP_ID_EXP (pxcap)
+ * PCI_CAP_ID_MSIX (msixcap)
+ * PCI_CAP_ID_PM (pmcap)
+ *
+ * However, they must still be explicitly initialized and added here.
  *
- * @returns 0 on success, -1 on failure and sets errno.
+ * FIXME: are we really making the user init these too? why? Is there a better
```

Comment:
AFAIK there's no other way around it, the capabilities must be configured by the user. We can introduce APIs to do this but I'm not sure how helpful that would be, capabilities are simple structs with many fields.